src/CameraZoomPlugin.cc

```diff
@@ -0,0 +1,342 @@
+/*
+   Copyright (C) 2023 ArduPilot.org
+
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "CameraZoomPlugin.hh"
+
+#include <atomic>
+#include <mutex>
+#include <string>
+
+#include <gz/common/Profiler.hh>
+
+#include <gz/math/Angle.hh>
+#include <gz/math/Pose3.hh>
+#include <gz/math/Quaternion.hh>
+
+#include <gz/plugin/Register.hh>
+
+#include <gz/rendering/Camera.hh>
+#include <gz/rendering/RenderEngine.hh>
+#include <gz/rendering/RenderingIface.hh>
+#include <gz/rendering/Scene.hh>
+
+#include <gz/sim/components/Camera.hh>
+#include <gz/sim/components/Model.hh>
+#include <gz/sim/components/Name.hh>
+#include <gz/sim/components/ParentEntity.hh>
+#include <gz/sim/components/Link.hh>
+#include <gz/sim/components/World.hh>
+#include <gz/sim/rendering/Events.hh>
+#include "gz/sim/Events.hh"
+#include <gz/sim/Link.hh>
+#include <gz/sim/Model.hh>
+#include <gz/sim/Sensor.hh>
+#include <gz/sim/World.hh>
+#include <gz/sim/Util.hh>
+
+#include <gz/transport/Node.hh>
+
+#include <sdf/Camera.hh>
+#include <sdf/Sensor.hh>
+
+namespace gz {
+namespace sim {
+inline namespace GZ_SIM_VERSION_NAMESPACE {
+namespace systems {
+
+//////////////////////////////////////////////////
+class CameraZoomPlugin::Impl
+{
+  /// \brief Handle a zoom command.
+  public: void OnZoom(const msgs::Double &_msg);
```

Comment:
Have you considered specifying zoom as a focal length? It's more deterministic.

Reply:
Yes, could do. This is just a basic first pass to get the internal mechanics of the plugin working and a factor applied to HFOV was the method used in the PX4-SITL_Gazebo plugin for Gazebo Classic so I took my steer from that.